source_mesh.py

```diff
@@ -0,0 +1,420 @@
+        [ids_wedge[0], ids_wedge[2], ids_wedge[4], ids_wedge[3]]
+    ]
+    # Create tetrahedra for wedge
+    for vertex_ids in wedge_canon_ids:
+        create_tet(
+            mbc, tag_handle, mesh_set, mbc_verts, verts, verts_s, vertex_ids,
+            strengths
+        )
+
+
+def create_mesh(
+    mbc, tag_handle, num_s, num_theta, num_phi, s_list, theta_list, phi_list,
+    mbc_verts, verts, verts_s):
+    """Creates volumetric source mesh in real space.
+
+    Arguments:
+        mbc (object): PyMOAB core instance.
+        tag_handle (TagHandle): PyMOAB source strength tag.
+        num_s (int): number of closed magnetic flux surfaces defining mesh.
+        num_theta (int): number of poloidal angles defining mesh.
+        num_phi (int): number of toroidal angles defining mesh.
+        s_list (list of float): list of closed flux surfaces defining mesh in
+            confinement space.
+        theta_list (list of float): list of poloidal angles defining mesh in
+            confinement space (rad).
+        phi_list (list of float): list of toroidal angles defining mesh in
+            confinement space (rad).
+        mbc_verts (list of EntityHandle): list of mesh vertices.
+        verts (list of list of float): list of 3D Cartesian coordinates of each
+            vertex in form [x (cm), y (cm), z (cm)].
+        verts_s (list of float): list of closed flux surface indices for each
+            vertex.
+            
+    Returns:
+        strengths (list of float): list of source strengths for each
+            tetrahedron (1/s).
+    """
+    # Instantiate tetrahedra mesh set in PyMOAB
+    tet_set = mbc.create_meshset()
+    # Add vertices to mesh set
+    mbc.add_entity(tet_set, mbc_verts)
+
+    # Initialize list of tetrahedra source strengths
+    strengths = []
+    
+    # Create tetrahedra, looping through vertices
+    for i, phi in enumerate(phi_list[:-1]):
+        # Create tetrahedra for wedges at center of plasma
+        for k, theta in enumerate(theta_list[:-1]):
+            # Define six wedge vertex indices in matrix format as well as index
+            # offset accounting for magnetic axis vertex in the form
+            # [toroidal index, flux surface index, poloidal index, offset]
+            wedge_id_data = [
+                [i,     0, 0,     i],
+                [i,     0, k,     i + 1],
+                [i,     0, k + 1, i + 1],
+                [i + 1, 0, 0,     i + 1],
+                [i + 1, 0, k,     i + 2],
+                [i + 1, 0, k + 1, i + 2]
+            ]
+            # Initialize list of vertex indices for wedges
+            ids_wedge = []
+            # Define vertex indices in row-major order for wedges
+            for vertex in wedge_id_data:
+                ids_wedge += [
+                    int(np.ravel_multi_index(
+                        vertex[:3], [num_phi, num_s - 1, num_theta]
+                    )) + vertex[3]
+                ]
```

Comment:
This seems more complicated than it needs to be... let's discuss in a software meeting. Maybe it's because the array dimensions are not consistent because of the magnetic axis?

Reply:
Yes, that's correct. I could change the structure so that there's a magnetic axis value for each theta angle (returns the same point) but that seems wasteful from a storage perspective

Reply:
Doing so would result in roughly `num_phi*num_theta` additional points